tools/testing/e2e/iree-e2e-conv2d-test.cc

```diff
@@ -36,57 +36,219 @@ static int convert_to_1d_index(iree_hal_dim_t channels, iree_hal_dim_t height,
+    // The layout of output tensor is NxOHxOWxf
+    iree_hal_dim_t out_idx =
+        convert_to_1d_index(oh_size, ow_size, f_size, n, oh, ow, oc);
 
-  float acc = acc_data ? iree_math_f16_to_f32(acc_data[out_idx]) : 0.f;
+    float acc = acc_data ? iree_math_f16_to_f32(acc_data[out_idx]) : 0.f;
 
-  for (iree_hal_dim_t ic = 0; ic < c_size; ++ic) {
     for (iree_hal_dim_t kh = 0; kh < kh_size; ++kh) {
       for (iree_hal_dim_t kw = 0; kw < kw_size; ++kw) {
-        iree_hal_dim_t inp_idx = convert_to_1d_index(
-            c_size, h_size, w_size, n, ic, (oh * sh_size + kh * dh_size),
-            (ow * sw_size + kw * dw_size));
-        iree_hal_dim_t krnl_idx =
-            convert_to_1d_index(c_size, kh_size, kw_size, oc, ic, kh, kw);
-
-        acc += iree_math_f16_to_f32(input_data[inp_idx]) *
-               iree_math_f16_to_f32(kernel_data[krnl_idx]);
+        for (iree_hal_dim_t ic = 0; ic < c_size; ++ic) {
+          iree_hal_dim_t inp_idx = convert_to_1d_index(
+              h_size, w_size, c_size, n, (oh * sh_size + kh * dh_size),
+              (ow * sw_size + kw * dw_size), ic);
+          iree_hal_dim_t krnl_idx =
+              convert_to_1d_index(kw_size, c_size, f_size, kh, kw, ic, oc);
+
+          acc += iree_math_f16_to_f32(input_data[inp_idx]) *
+                 iree_math_f16_to_f32(kernel_data[krnl_idx]);
+        }
       }
     }
     result_data[out_idx] = iree_math_f32_to_f16(acc);
   }
 }
 
-static void reference_conv2d_f32_f32_f32_f32(
+// [f32 <= f16 * f16 + f32]
+static void reference_conv2d_f16_f16_f32_f32(
+    iree_hal_dim_t n_size, iree_hal_dim_t c_size, iree_hal_dim_t h_size,
+    iree_hal_dim_t w_size, iree_hal_dim_t f_size, iree_hal_dim_t kh_size,
+    iree_hal_dim_t kw_size, iree_hal_dim_t layout, iree_hal_dim_t sh_size,
+    iree_hal_dim_t sw_size, iree_hal_dim_t dh_size, iree_hal_dim_t dw_size,
+    iree_hal_dim_t oh_size, iree_hal_dim_t ow_size, const uint16_t* input_data,
+    const uint16_t* kernel_data, const float* acc_data, float* result_data,
+    iree_hal_dim_t n, iree_hal_dim_t oc, iree_hal_dim_t oh, iree_hal_dim_t ow) {
+  if (layout == 0) {
+    // The layout of output tensor is NxfxOHxOW
+    iree_hal_dim_t out_idx =
+        convert_to_1d_index(f_size, oh_size, ow_size, n, oc, oh, ow);
+
+    float acc = acc_data ? acc_data[out_idx] : 0.f;
+
+    for (iree_hal_dim_t ic = 0; ic < c_size; ++ic) {
+      for (iree_hal_dim_t kh = 0; kh < kh_size; ++kh) {
+        for (iree_hal_dim_t kw = 0; kw < kw_size; ++kw) {
+          iree_hal_dim_t inp_idx = convert_to_1d_index(
+              c_size, h_size, w_size, n, ic, (oh * sh_size + kh * dh_size),
+              (ow * sw_size + kw * dw_size));
+          iree_hal_dim_t krnl_idx =
+              convert_to_1d_index(c_size, kh_size, kw_size, oc, ic, kh, kw);
+
+          acc += iree_math_f16_to_f32(input_data[inp_idx]) *
+                 iree_math_f16_to_f32(kernel_data[krnl_idx]);
+        }
+      }
+      result_data[out_idx] = acc;
```

Comment:
All this math in C++ scares me, especially for lower precision types. Would like to at least validate the C++ against reference python framework code _then_ compare our own compiler/runtime/codegen systems against the C++ once we trust it